### DISCUSSION TITLE
WSLG doesn't release unused GPUs and doesnt autoshutdown

### DISCUSSION THREAD

**Discussion by hemanth981** [microsoft/wslg#492] [Q&A] [answered]
(Sorry that this is a bit long)
So I've been trying to get GUI based linux applications to run off of WSLG on a device with multiple GPUs (Intel and Nvidia). GPU selection is an important parameter for me, since the overhead on using the dedicated GPU is much higher than using the integrated on on WSL. Many discussions suggested setting the MESA_D3D12_ADAPTER_NAME environment variable, which didn't work out, but then I stumbled on the latest github commit in mesa repo on this issue in which the environment variable used was MESA_D3D12_DEFAULT_ADAPTER_NAME . So I finally got the GUI apps to run off of the Intel GPU, but there's a problem:
Even if I'm using only the Intel GPU, on starting any GUI apps from wslg, there seems to be a process named 'System' which starts running on the dedicated GPU also (can see this from Nvidia GPU activity monitor on Windows). I'm quite sure the app is running on the intel GPU, as glxinfo -B gives the renderer as Intel. Also, on running any linux GUI app on the dedicated Nvidia GPU, the Nvidia GPU activity monitor displays one process named 'System' for each GUI process running, along with one additional 'System' process (which I think is some sort of keepalive to keep the GPU active?) which starts with the first GUI app launched.
So I'm guessing MESA enumerated the GPUs to find the searchstring "INTEL" to use the right GPU, but then the Nvidia GPU was never released. 
Also, it seems that as long as the 'System' process shows up in the GPU activity monitor, WSL never auto shuts down. Or, if I had launched any GUI app instance in WSL, the Nvidia GPU is kicked active and neither WSL, nor the dedicated GPU becomes inactive until I manually shutdown WSL.

So here are the issue in short:
1) Dedicated GPU also becomes active when starting any GUI app on the integrated GPU in WSLg
2) Dedicated GPU is never released (always active) even when not in use by WSL, until WSL is manually shutdown.
2) WSL doesn't auto shutdown if any GUI app was started at least once after starting WSL

Is this expected behaviour, or is it something wrong with my installation?
(I'm running Windows 11 Home 22000.194 , with all updates for WSL installed using wsl --update)

**Comment by hemanth981** [comment, marked as answer]
I think I found a solution: force software rendering in WSL system distro by adding LIBGL_ALWAYS_SOFTWARE=1 in .wslgconfig 
So I assume the system distro is connecting to and acting as a keepalive for the GPU, eventhough the renderer actually being used by my user distro is the intel one. (Maybe it keeps alive all the GPUs it enumerated?)
(Wouldn't it be better if it kept alive only the preferred renderer/the one last used, rather than all the GPUs?)

Also, now I can start gui apps from within the wsl terminal using the intel GPU and not kick start the Nvidia one after exporting MESA_D3D12_DEFAULT_ADAPTER_NAME="INTEL" . I don't know if this is actually a good solution, or what all other implications this have.
I still haven't gotten this to work in starting the apps from the start menu, as they only seem to run on the dedicated GPU. 
I can see from the properties of the shortcuts in the start menu that they are run using `wslg.exe -d <distroname> command`, and can manually run the apps on the required gpu using `wslg.exe -d <distroname> MESA_D3D12_DEFAULT_ADAPTER_NAME="INTEL" command` I'm not really sure how wslg.exe starts the GUI apps (which shell it uses), or if there's any way to pass `MESA_D3D12_DEFAULT_ADAPTER_NAME="INTEL"` as an environment variable without modifying all the shortcuts manually (which will be overwritten anyway), so can anyone please help with that?

Also, it would be really nice to have this setting (MESA_D3D12_DEFAULT_ADAPTER_NAME) in .wslconfig and get it to work.

**Comment by hemanth981** [comment]
Found how to make it work for apps launched from start menu too. You have to modify the profile file for your shell and add `export MESA_D3D12_DEFAULT_ADAPTER_NAME="INTEL"` in it. I was trying to modify .zshrc and .bashrc which didn't work, but changing .bash_profile (if bash) or .zprofile (if zsh) worked.

**Comment by McBride-SallyJoeBob** [comment]
I have a similar problem. My GPU does not get released either but even worse, I cant update my Mesa higher than 20.3.4 so I am stuck with software rendering. Even so, I too get that "System" process on my GPU and it does not get released until I shutdown WSL. Its such a pain because my Linux gui apps are not only consuming lots of power but are also still just using software rendering via llvmpipe

**Reply by hemanth981** [reply]
Did you try to force software rendering on the system distro by making .wslgconfig file and adding `LIBGL_ALWAYS_SOFTWARE=1` ? This might help in the releasing part.
Also, you wouldn't be stuck on sw rendering with older version of mesa, just that you can't manually choose which GPU to use, if you have multiple GPUs. Try to follow the instructions on wslg page (making sure you are on the right build, updating wsl using wsl --update, etc)

**Reply by McBride-SallyJoeBob** [reply]
I do have two gpus: amd integrated and an RTX 2060. Adding the .wslgconfing file and adding the system distro environments worked! My gpu gets released. I still don't know how to solve the llvmpipe issue. I tried upgrading the Mesa driver to the latest version. Even though it says that I have Mesa 21 installed, glxinfo says I have mesa 20.3.4. Wsl is updated to the the latest version and i am in windows 11 22000.194. Additionally, the dev/dxg file exists. I am so confused
![image](https://user-images.githubusercontent.com/39777219/137030911-9ea7a430-7300-4a83-9189-75d44efc5e44.png)

**Reply by hemanth981** [reply]
You may want to try this: https://launchpad.net/~kisak/+archive/ubuntu/kisak-mesa
It worked for me, but I think it should be actually working even with the older mesa, just that you wont have the option to select which GPU you want.
I'm not really sure how to solve that, but try this and check (inside your user distro):
export LIBGL_ALWAYS_SOFTWARE=0
export MESA_D3D12_DEFAULT_ADAPTER_NAME="NVIDIA"
unset GALLIUM_DRIVER
glxinfo -B

Also, make sure you can see your GPUs inside your WSL user distro (Check the output of lspci , or look for the string 'display' in the output of lshw)

**Reply by McBride-SallyJoeBob** [reply]
lshw shows two Microsoft GPUS with driver dxgkrnl. One of them is on display 0 and the other is on display 1. Ive tried everything when updating mesa using the PPA. I used ppa purge and even tried the experimental versions. Even though it says it is upgrading to version 21,  glxinfo remains unchanged and stuck on version 20.3.4. I am unsure whether this is relevant but I get a message that  /usr/lib/wsl/lib/libcuda.so.1 is not a symbolic link when i tried to upgrade the ppa. Also using the commands you listed for export and unset made no difference either ☹️

` *-display:0
          description: 3D controller
          product: Microsoft Corporation
          vendor: Microsoft Corporation
          physical id: 2
          bus info: pci@1984:00:00.0
          version: 00
          width: 32 bits
          clock: 33MHz
          capabilities: bus_master cap_list
          configuration: driver=dxgkrnl latency=0
          resources: irq:0
`
`*-display:1
          description: 3D controller
          product: Microsoft Corporation
          vendor: Microsoft Corporation
          physical id: 4
          bus info: pci@364d:00:00.0
          version: 00
          width: 32 bits
          clock: 33MHz
          capabilities: bus_master cap_list
          configuration: driver=dxgkrnl latency=0
          resources: irq:0
`

**Reply by McBride-SallyJoeBob** [reply]
![image](https://user-images.githubusercontent.com/39777219/137167058-a99ad63c-054c-4b60-a26e-21e3ab1e9df8.png)
![image](https://user-images.githubusercontent.com/39777219/137167185-111b93b7-729b-43cc-a8d2-def96709637f.png)

here are my dxdiags to see if there are any driver issues

**Comment by hemanth981** [comment]
Seems like your GPUs are detected, but wsl isn't using it at all... Seems really strange. 
Just to make sure, can you try the following steps:
Open an elevated powershel and run `wsl --update`
Update Nvidia/amd drivers to their latest version and test
Also, if you are using Ubuntu 18, try installing 20 (if it still doesnt work, then try reinstalling your distro if you have nothing too important within the distro/after taking a backup. Maybe a clean start might fix things up)
There was some troubleshoot guide for wslg somewhere on github, but i can't seem to find it right now. I'll add a link to that if I stumble across it again.

**Reply by McBride-SallyJoeBob** [reply]
I am using ubuntu 20. I can run the CUDA compute examples using the NVIDIA docker no problem. Its just specifically for openGL rendering, the problem exists. Hopefully someone here can figure out what the issue might be since more and more people will be using the software in the future.

**Comment by McBride-SallyJoeBob** [comment]
I found the Problem!!! There was a dependency on my installation AMDGPU driver which prevented the newest LibGL mesa version from being used. I can now access the hardware accelarated GPUS. However, my performance is strangly bad 30fps on glxgears

**Reply by hemanth981** [reply]
Congrats!!! Glad that you solved it. The fps values in glxgears is not that good for me too (100fps for NVIDIA GPU, 160 fps for Integrated GPU, and some insane 700fps on CPU based rendering), but on using some heavy applications that requires hardware rendering, you can see the difference. Maybe this performance issue is due to the additional overhead required to transfer data from the gpu memory to system ram so that wsl can access it, but on running heavy rendering applications, the bottleneck more or less shifts to the GPU processing power rather than memory bandwidth.

**Reply by McBride-SallyJoeBob** [reply]
Yeah my fps is really poor though even worse than yours. its 30 for the integrated and 12 for the NVIDIA gpu. Gui apps don't feel graphically snappier at all compared to software rendering. Your probably right there is some bottleneck. I still cant fix the gpu release problem. So i will be stuck with llvmpipe rendering still until I get an answer to this issue. I cant be using wslg with my NVIDIA gpu idling because it will drain my battery

**Reply by hemanth981** [reply]
Actually, I found a workaround for the GPU release problem which I had mentioned in my answer. The problem is, whenever you use any app that requires GPU over wslg, the system distro engages the (all the) GPU(s) and doesn't release them. You can force the system distro to use software rendering (this didn't break anything, at least for me) by adding
```LIBGL_ALWAYS_SOFTWARE=1```
 in .wslgconfig
This will make sure the system distro uses software rendering and won't engage the dedicated GPU, while the apps you run on the user distro can use the GPU if necessary like usual.

**Reply by luciandf** [reply]
> I found the Problem!!! There was a dependency on my installation AMDGPU driver which prevented the newest LibGL mesa version from being used. I can now access the hardware accelarated GPUS. However, my performance is strangly bad 30fps on glxgears

do you mind explaining a bit more what was the dependency in AMDGPU driver?
